Add wp-cli integration

matiasbenedetto opened the pull request:
## What?
Add wp-cli integration for theme-check plugin

## Why?
To use theme-coutput of the theme-check plugin in automated envs like CI workflows.

## How to use and how to test
- In your WordPress install `theme-check` plugin **using this branch**.
- Install `wp-cli` in your system.
- Run in your console `wp theme-check run`.
- Observe the output in your console

It should look like this:

Running `wp theme-check run` :
![image](https://github.com/user-attachments/assets/139f53e4-e73a-45f5-9b1c-ccec1275f603)


Running `wp theme-check run --format=json`:
![image](https://github.com/user-attachments/assets/d939eaaa-e903-45ab-ba8f-aaf34d9029c1)



## Screencast
[Screencast from 06-09-24 11:39:55.webm](https://github.com/user-attachments/assets/0add5f81-c3c0-40ee-89f0-ab0e9ea88fde)

jffng commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+			$this->display_themechecks_as_json( $check_theme_slug );
+		} else {
+			if ( ! $success ) {
+				WP_CLI::error( "Theme check failed for {$slug}." );
```
```suggestion
				WP_CLI::error( "Theme check failed for {$check_theme_slug}." );
```

jffng commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+			if ( ! $success ) {
+				$json_output = array(
+					'check-completed' => false,
+					'result'          => "Error: Theme check failed for {$slug}.",
```
```suggestion
					'result'          => "Error: Theme check failed for {$check_theme_slug}.",
```

jffng reviewed (commented):
When the theme check fails, the error messages are not logged. All that is returned is: 

```
$ wp theme-check run
Error: Theme check failed for twentytwentyfive.
```

jffng commented:
Nice change! I can't push to your branch, but here is some suggestion changes to the readme:


### Usage with wp-cli

To use with [wp-cli](https://wp-cli.org/), ensure the theme check plugin is active and `wp-cli` is installed. The `theme-check` subcommand is added to `wp-cli` and can be used as follows:

`wp theme-check run [<theme>] [--format=<format>]`

#### Options
| Option | Accepts | Required | Default
| -- | -- | -- | -- | 
| `theme` | The slug of the theme to check | No | Current theme slug
| `format` | `cli` or `json` | No | `cli`

#### Examples
`wp theme-check run`
`wp theme-check run twentytwentyfour`
`wp theme-check run --format=json`
`wp theme-check run twentytwentyfour --format=json`

matiasbenedetto commented:
@jffng cool, look good. I added you as a collaborator in the fork so you can push :)

codestylist commented:
What a cool feature @matiasbenedetto . Thank you for this! Hope it will be in the master soon!

ernilambar commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+	 * [<theme>]
+	 * : The slug of the theme to check. If not provided, checks the current theme.
+	 *
+	 * [--format=<format>]
```
If you follow this format https://github.com/wp-cli/entity-command/blob/main/src/Post_Command.php#L419 then WP-CLI itself will validate the available options. That way you won't need `in_array()` check in the method.

matiasbenedetto replied:
If I removed that snippet:
```
		if ( ! in_array( $format, array( 'cli', 'json' ) ) ) {
			WP_CLI::error( "Invalid format. Accepts 'cli' or 'json'." );
			return;
		}
```
If I use a non-existent format the function runs using the default format instead of raising an error. No matter the comment format implemented in the function. Is that expected?

matiasbenedetto replied:
It's not working like that on my end. Removing that snippet and using a random format doesn't throw any error:

![image](https://github.com/user-attachments/assets/a67dab26-28a4-4e67-a9a2-ec3792a9d8d5)

matiasbenedetto replied:
It's working fine after implementing your comment formatting suggestion (https://github.com/WordPress/theme-check/pull/458#discussion_r1750604286), thanks!

ernilambar commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+				return;
+			}
+			WP_CLI::error( "Theme '{$check_theme_slug}' not found." );
+			return;
```
WP_CLI::error() exits the process. So I don't think `return` is needed.

matiasbenedetto replied:
:+1: removed here: https://github.com/WordPress/theme-check/pull/458/commits/37542b38dd2ac7b8d2125f67dcc4cfc59dfa4541

ernilambar commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+					'result'          => "Error: Theme check failed for {$check_theme_slug}.",
+					'messages'        => array(),
+				);
+				WP_CLI::log( wp_json_encode( $json_output, JSON_PRETTY_PRINT ) );
```
Note that output of `log()` is discarded to STDOUT when `--quiet` flag is passed. But here I think we should display JSON output whatever case if quiet flag is present or not.

matiasbenedetto replied:
Ok, I used `line()` to log the results. 

I found 2 issues regarding wp-cli loggers while working on this:
- https://github.com/wp-cli/wp-cli/pull/5979
- https://github.com/wp-cli/wp-cli/issues/5980

vcanales replied:
> But here I think we should display JSON output whatever case if quiet flag is present or not.

Not sure I agree with bypassing `--quiet`, I would expect no output at all. An alternative is to make sure that the program always exits with a descriptive code; ie. 0 for OK, non-0 for errors.

edit: On reading @matiasbenedetto's comment below, seems like the checks are not meant to return a boolean.

> It's a little bit counterintuitive, but the theme-check plugin doesn't work with the objective of outputting a boolean ("theme OK/ theme with errors"). The plugin is not designed like that. It just outputs a list of messages regarding the theme, but it never says whether this theme is OK or not. Maybe in the future, we could implement that, but that's not the case yet.

Still, I don't think bypassing `--quiet` is cool.

matiasbenedetto replied:
I think this was fixed on the latest commits.

ernilambar commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,179 @@
+	 * [--format=<format>]
+	 * : Output format. Accepts 'cli' or 'json'. Default: 'cli'.
+	 *
+	 * ## EXAMPLES
```
Recommended to follow this format with short explanation of the command. https://github.com/wp-cli/entity-command/blob/main/src/Post_Command.php#L467

matiasbenedetto replied:
done here: https://github.com/WordPress/theme-check/pull/458/commits/75c2f9cf123ebb6ec2272436cf421429dd6ad7d8

ernilambar commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,157 @@
+	/**
+	 * Run a theme check on the specified theme or the current theme.
+	 *
+	 * @param array $args       Indexed array of positional arguments.
```
These 3 lines are breaking the expected doc format of WP-CLI command. So format validation is not working. You can confirm it by checking `wp help theme-check run`. Currently you wont see arguments properly. When you remove these lines, it will work as expected.

ernilambar replied:
You can move these after Examples.

matiasbenedetto replied:
Good. Done here: https://github.com/WordPress/theme-check/pull/458/commits/e998dc02035639c6f0bb3fa297842a09391abd95

ernilambar commented on .vscode/launch.json:
```diff
@@ -0,0 +1,48 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
May be we need to update gitignore to ignore this type of IDE files and folders.

matiasbenedetto replied:
I commited that file accidentally. I removed it here: https://github.com/WordPress/theme-check/pull/458/commits/8d61a5995bed77cdcf42dad6259e7549bb0f61f4

I will update the .gitignore file in another PR.

matiasbenedetto replied:
Submitted a PR for the gitignore: https://github.com/WordPress/theme-check/pull/460

ernilambar commented:
When I run check on a theme with errors. `wp theme-check run mazebox`. Output is:

```
Error: Theme check failed for mazebox.
```

But when there is no errors, output includes recommendation and info items.

```
Success: Theme check completed for mazebox.

RECOMMENDED: No reference to register_block_pattern was found in the theme. Theme authors are encouraged to implement custom block patterns as a transition to block themes.

RECOMMENDED: No reference to register_block_style was found in the theme. Theme authors are encouraged to implement new block styles as a transition to block themes.
...

```

matiasbenedetto commented:
> When I run check on a theme with errors. wp theme-check run mazebox. Output is:
> Error: Theme check failed for mazebox.

I don't know what kind of errors are you checking but I think this is expected. Theme-check will return an error message if the check couldn't be run for some reason (errors parsing theme code for example). The success and error messages are only used to signal if the plugin was able to run the checks or not, not to output the results of those checks.

> But when there is no errors, output includes recommendation and info items.

It's a little bit counterintuitive, but the theme-check plugin doesn't work with the objective of outputting a boolean ("theme OK/ theme with errors"). The plugin is not designed like that. It just outputs a list of messages regarding the theme, but it never says whether this theme is OK or not. Maybe in the future, we could implement that, but that's not the case yet.

ernilambar commented:
Example:

```
$ wp theme install twentysixteen --activate
Installing Twenty Sixteen (3.3)
Downloading installation package from https://downloads.wordpress.org/theme/twentysixteen.3.3.zip...
Using cached file '/Users/nilambarsharma/.wp-cli/cache/theme/twentysixteen-3.3.zip'...
Unpacking the package...
Installing the theme...
Theme installed successfully.
Activating 'twentysixteen'...
Success: Switched to 'Twenty Sixteen' theme.
Success: Installed 1 of 1 themes.

$ wp theme install twentyten --activate --quiet
$ echo $?
0
```

Another case:

```
$ wp theme list --format=json --quiet
[{"name":"twentytwelve","status":"inactive","update":"none","version":"4.3","update_version":"","auto_update":"off"}]

$ wp theme list --format=json
[{"name":"twentytwelve","status":"inactive","update":"none","version":"4.3","update_version":"","auto_update":"off"}]
```

There is no right or wrong approach. We can go for any approach we agree upon. But from my experience, general expectation in CLI is that if I pass `--format=json` then I would expect output in valid JSON format, whether quiet flag is passed or not.

vcanales commented:
> There is no right or wrong approach. We can go for any approach we agree upon. But from my experience, general expectation in CLI is that if I pass `--format=json` then I would expect output in valid JSON format, whether quiet flag is passed or not.

I see now. Yes, I agree that if we're explicitly asking for a format, it makes sense to bypass `--quiet`. On the other hand, these seem like incompatible flags to me; why would someone use both? This is an unrelated discussion, though.

jffng commented on wp-cli/class-theme-check-cli.php:
```diff
@@ -0,0 +1,160 @@
+			if ( ! $success ) {
+				WP_CLI::error( "Theme check failed for {$check_theme_slug}." );
+			}
+			$this->display_themechecks_in_cli( $check_theme_slug );
```
Execution stops if a `WP_CLI::error` is thrown, so the errors are not being displayed. Could this line be moved above the $success check?

matiasbenedetto replied:
I'm not sure, i'm seeing this logic here:
```php
// If the check process failed (if the checks were not able to run properly) raise an error. There's nothing else to do.
if ( ! $success ) {
WP_CLI::error( "Theme check failed for {$check_theme_slug}." );
}
// If the checking process run print in the console the checks. This will also print the things marked as an error in the theme.
$this->display_themechecks_in_cli( $check_theme_slug );
```
Could you explain why you think we should move it?

jffng replied:
```
// If the check process failed (if the checks were not able to run properly) raise an error. There's nothing else to do.
```

I don't think this is correct. `$success` is the result of `run_themechecks_against_theme` which returns `false` if the theme fails a check. 

> Could you explain why you think we should move it?

Using `twentytwentyfive`, here is the output currently: 

```
Error: Theme check failed for twentytwentyfive.
```

Moving call to `display_themechecks_in_cli` above the conditional, the result is:

```
Success: Theme check completed for twentytwentyfive.

REQUIRED: phpcs.xml.dist XML file found. This file must not be in the production version of the theme.

WARNING: assets/images/image-from-rawpixel-id-2222755.webp is 997.6 KB in size. Large file sizes have a negative impact on website performance and loading time. Compress images before using them.

WARNING: assets/images/image-from-rawpixel-id-2224378.webp is 879.6 KB in size. Large file sizes have a negative impact on website performance and loading time. Compress images before using them.

WARNING: assets/images/location.webp is 778.1 KB in size. Large file sizes have a negative impact on website performance and loading time. Compress images before using them.

WARNING: assets/images/poster-image-background.webp is 574.7 KB in size. Large file sizes have a negative impact on website performance and loading time. Compress images before using them.

INFO Themes that use the tag accessibility-ready will need to undergo an accessibility review. See https://make.wordpress.org/themes/handbook/review/accessibility/

WARNING: Your theme appears to be in the wrong directory for the theme name. The directory name must match the slug of the theme. This theme's correct slug and text-domain is twenty-twenty-five. (If this is a child theme, you can ignore this error.)

INFO: Only one text-domain is being used in this theme. Make sure it matches the theme's slug correctly so that the theme will be compatible with WordPress.org language packs. The domain found is twentytwentyfive.

RECOMMENDED: Requires PHP is recommended to have major and minor versions only (e.g. 7.4). Patch version is not needed (e.g. 7.4.1).

Error: Theme check failed for twentytwentyfive.
```

I think the latter is more useful because it reports what checks failed (the first one marked `REQUIRED`). 

Also I am not sure we should throw an error here if `$success` isn't true, because the theme check successfully ran, but the result was it failed the checks.

ernilambar replied:
This command's focus should be on pass/fail check for a theme, rather than run is successfully completed or not. And exit code should be returned based on that I believe.

matiasbenedetto replied:
I think I could implement your feedback, please see: https://github.com/WordPress/theme-check/pull/458#issuecomment-2349033538

ernilambar commented:
Just an idea: What if we simply list check results like CLI displays. Below is one example. 
Each message is splitted into `type` and `message`. This way we could support different format using WP_CLI\Formatter(), like `table`, `json`, `csv`, `yaml`.

When I need to check in my CI runner where I just need to check `required` items or not, I could use:

```
wp theme-check run theme-slug --type=error --format=json
```

<img width="900" alt="Screenshot 2024-09-11 at 11 18 34 PM" src="https://github.com/user-attachments/assets/be6d057d-0b61-43d0-900f-70f6874f4fb9">

matiasbenedetto commented:
I went through the theme-check plugin code again, and I found that we can get a boolean signaling if the theme has required changes or not. Knowing that I was able to I implemented all the latest feedback on the last commit:

- Exit code depending on the existence of required changes or not. https://github.com/matiasbenedetto/theme-check/blob/d82740ea74277dd3e776019594ab10923cfbdd60/wp-cli/class-theme-check-cli.php#L86
- Simpler code
- Use the wp-cli table formatting by default. 

These are the results using TT4 theme (I removed `theme.json` file and screenshot image to get REQUIRED changes):

```
wp theme-check run
```
![image](https://github.com/user-attachments/assets/3e8182df-c7e9-4ff4-b9f7-e2ba7d3c4a7a)

---
```
wp theme-check run --format=json
```
![image](https://github.com/user-attachments/assets/c200f79a-51bd-4d7b-90af-66861e2ced3f)


---
```
wp theme-check run non-existing-theme
```
![image](https://github.com/user-attachments/assets/0b540543-f75e-455d-9633-7576468f6da2)

jffng commented on README.md:
```diff
@@ -50,5 +50,23 @@ add_filter( 'tc_skip_development_directories', '__return_true' );
+To use with [wp-cli](https://wp-cli.org/), ensure the theme check plugin is active and `wp-cli` is installed. The `theme-check` subcommand is added to `wp-cli` and can be used as follows:
+
+`wp theme-check run [<theme>] [--format=<format>]`
+
```
```suggestion

On success, the command returns a formatted table of results from the theme check plugin.

```

jffng reviewed (approved):
It works well in my testing.

One question that does not have to be a blocker, but do you know how to add the help text for the `theme-check` subcommand when a user runs `wp help`?

![Screenshot 2024-09-16 at 4 35 38 PM](https://github.com/user-attachments/assets/926dcae8-df20-40b2-a79c-01d0da401882)

matiasbenedetto commented:
> One question that does not have to be a blocker, but do you know how to add the help text for the theme-check subcommand when a user runs wp help?

Added it here: https://github.com/WordPress/theme-check/pull/458/commits/d60d1e3c4d563d06d3d5a4fe721d05afd23ac6ed

![Screenshot from 2024-09-16 17-57-05](https://github.com/user-attachments/assets/00ac3a3d-f180-41b9-8145-735fb2163910)
